### ISSUE TITLE
Parquet Hadoop parser fails to parse columns specified in transformSpec only.

### ISSUE THREAD

**Issue by zhiyuany** [apache/incubator-druid#7169]
# Affected Version

druid-0.13.0-incubating

# Description

**The scene** 
We have data stored in hdfs with parquet format, named `p-event`.
Each row in the data has a column `execute-timestamp`, Long or Null indicating execution time.
Now we want a metric `execution-count` to count how many of the p-event has been executed.

We use below indexing-spec:
```
{
  "type": "index_hadoop",
  "spec": {
    ...
    "dataSchema": {
      ...,
      "parser": {
        "type": "parquet",
        "parseSpec": {
          "format": "timeAndDims",
          "timestampSpec": ...,
          "dimensionsSpec": {
            "dimensions": [
              ...
            ],
            "dimensionExclusions": [],
            "spatialDimensions": []
          }
        }
      },
      "metricsSpec": [
        ...
        {
          "type" : "longSum",
          "name" : "execution-count",
          "fieldName" : "has-execution"
        }
      ],
      "transformSpec": {
        "transforms": [
          {
            "type": "expression",
            "name": "has-execution",
            "expression": "if (execute-timestamp > 0, 1, 0)"
          }
        ]
      }
    }
  }
}

```
_Note that `execute-timestamp` is not part of the `dimensions`, only declared in `transform`._

**The bug**
_**`execution-count` is constantly 0**_
The data has tons of event with `execute-timestamp` satisfying the predicate  _"if (execute-timestamp > 0, 1, 0)"_, but metric `execution-count` is constantly 0, as if the column `execute-timestamp` is always null/absent.
The hypothesis is that column `execute-timestamp` is not parsed. 

**The reproduce**
Above scene walk through provide the way to reproduce the problem. 

**How to unbroke/fix it**
***A***
One fix to the problem is _adding `execute-timestamp` to the indexing-spec `dimensions` section_, and this leads to "execution-count" produce correct metric.
But apparently this is not the right use case, and output would explode in size when we put high cardinality fields like timestamp to dimension.
```
{
  ...
          "dimensionsSpec": {
            "dimensions": [
              "execute-timestamp"
            ],
            ...
          }
        }
      },
  ...
}
```
***B***
Same as solution A, _adding `execute-timestamp` to the indexing-spec `metrics` section_ has same effect.
```
{
  ...
          "metricsSpec": [
             {
               "type" : "longSum",
               "name" : "dummy-metric",
               "fieldName" : "execute-timestamp"
             }
            ]
            ...
          }
        }
      },
  ...
}
```

***C***
Another hack fix to the problem is _transform the field in place_ or _shadow_ .
But this remove the original value of `execute-timestamp` and disallow other transformation.
```
{
  ...
      "metricsSpec": [
        ...
        {
          "type" : "longSum",
          "name" : "execution-count",
          "fieldName" : "execute-timestamp"
        }
      ],
      "transformSpec": {
        "transforms": [
          {
            "type": "expression",
            "name": "execute-timestamp",
            "expression": "if (execute-timestamp > 0, 1, 0)"
          }
        ]
      }
  ...
}

```

**Debugging**
There is no error found in the indexing job, besides the result is inaccurate. 
No further debugging performed.

**Comment by clintropolis** [comment]
I believe this is an artifact of how the "contrib" Parquet extension functions, [that it only reads columns from the file that are specified in the schema.](https://github.com/apache/incubator-druid/blob/cf15aac71f9f0f2ba41a2304d6b72719e6f7ec69/extensions-contrib/parquet-extensions/src/main/java/org/apache/parquet/avro/DruidParquetReadSupport.java#L57)

Fwiw, in the upcoming Druid 0.14 the Parquet extension has been reworked and moved to a "core" extension (see #6360). In the new version of the extension, there are added `parquet` and `parquet-avro` parser types which support a `flattenSpec`, and _also_ allow a `transformSpec` to refer to columns which are not otherwise used as dimensions or metrics. Note that using the `timeAndDims` parser type will still run into this same issue, as the optimization to use a partial schema to read the file is still used in this case. I tested to confirm this behavior, both that using `parquet` and `parquet-avro` with 'auto discovery' enabled allow a transform expression to refer to a column that isn't a dimension or metric, and that using `timeAndDims` will experience the current behavior that you are running into.

**Comment by gianm** [comment]
Based on @clintropolis's comment it looks like this is going to be fixed in 0.14.0.
